File: typescript/core/test/cron-occurrences.test.ts
import { readFile } from "node:fs/promises";
import { describe, expect, it } from "vitest";
import { dueOccurrences } from "../src/worker.js";

interface CronOccurrenceFixture {
  id: string;
  expression: string;
  timezone: string;
  lastOccurrenceAt: string | null;
  now: string;
  limit: number;
  expected: string[];
}

const fixtures = JSON.parse(
  await readFile(new URL("../../../protocol/v1/cron-occurrences.json", import.meta.url), "utf8"),
) as CronOccurrenceFixture[];

describe("cron occurrence compatibility", () => {
  it.each(fixtures)("matches the shared table for $id", (fixture) => {
    const actual = dueOccurrences(
      fixture.expression,
      fixture.lastOccurrenceAt === null ? null : new Date(fixture.lastOccurrenceAt),
      new Date(fixture.now),
      fixture.limit,
      fixture.timezone,
    );

    expect(actual.map((occurrence) => occurrence.toISOString())).toEqual(
      fixture.expected.map((occurrence) => new Date(occurrence).toISOString()),
    );
  });
});

File: go/cron_occurrences_test.go
package workhorse

import (
	"encoding/json"
	"os"
	"testing"
	"time"
)

type cronOccurrenceFixture struct {
	ID               string   `json:"id"`
	Expression       string   `json:"expression"`
	Timezone         string   `json:"timezone"`
	LastOccurrenceAt *string  `json:"lastOccurrenceAt"`
	Now              string   `json:"now"`
	Limit            int      `json:"limit"`
	Expected         []string `json:"expected"`
}

func TestDueOccurrencesMatchTheSharedCronTable(t *testing.T) {
	contents, err := os.ReadFile("../protocol/v1/cron-occurrences.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixtures []cronOccurrenceFixture
	if err := json.Unmarshal(contents, &fixtures); err != nil {
		t.Fatal(err)
	}

	for _, fixture := range fixtures {
		t.Run(fixture.ID, func(t *testing.T) {
			location, err := time.LoadLocation(fixture.Timezone)
			if err != nil {
				t.Fatal(err)
			}
			now, err := time.Parse(time.RFC3339, fixture.Now)
			if err != nil {
				t.Fatal(err)
			}
			var lastOccurrence *time.Time
			if fixture.LastOccurrenceAt != nil {
				parsed, err := time.Parse(time.RFC3339, *fixture.LastOccurrenceAt)
				if err != nil {
					t.Fatal(err)
				}
				lastOccurrence = &parsed
			}

			actual, err := dueOccurrences(
				fixture.Expression,
				lastOccurrence,
				now,
				fixture.Limit,
				location,
			)
			if err != nil {
				t.Fatal(err)
			}
			if len(actual) != len(fixture.Expected) {
				t.Fatalf("expected %d occurrences, received %v", len(fixture.Expected), actual)
			}
			for index, occurrence := range actual {
				if occurrence.UTC().Format(time.RFC3339) != fixture.Expected[index] {
					t.Fatalf("occurrence %d: expected %s, received %s", index, fixture.Expected[index], occurrence.UTC().Format(time.RFC3339))
				}
			}
		})
	}
}
